svc/job/eta.go

```diff
@@ -0,0 +1,120 @@
+	"github.com/openmarketplaceengine/geoservices/distance"
+	"github.com/openmarketplaceengine/geoservices/distance/google"
+	"github.com/openmarketplaceengine/openmarketplaceengine/cfg"
+	"github.com/openmarketplaceengine/openmarketplaceengine/dao"
+	"github.com/openmarketplaceengine/openmarketplaceengine/dom/job"
+	"github.com/openmarketplaceengine/openmarketplaceengine/svc/location"
+	"googlemaps.github.io/maps"
+)
+
+type Estimate struct {
+	DistanceMeters int
+	Duration       time.Duration
+}
+
+type EstimatedJob struct {
+	ID              dao.SUID
+	WorkerToPickup  Estimate
+	PickupToDropOff Estimate
+	WorkerLocation  Location
+	PickupLocation  Location
+	DropOffLocation Location
+}
+
+type Location struct {
+	Address string
+	Lat     float64
+	Lng     float64
+}
+
+func estimatedJobs(ctx context.Context, workerLocation *location.LastLocation, jobs []*job.AvailableJob) ([]*EstimatedJob, error) {
+	if jobs == nil {
+		return nil, fmt.Errorf("jobs argument cannot be nil")
+	}
+	if len(jobs) == 0 {
+		return nil, fmt.Errorf("jobs argument cannot be empty")
+	}
+
+	client, err := maps.NewClient(maps.WithAPIKey(cfg.Server.GoogleAPIKey))
+	if err != nil {
+		return nil, fmt.Errorf("maps client create error: %w", err)
+	}
+
+	origins, destinations := transform(jobs)
+	originsPlus := append(origins, geoservices.LatLng{Lat: workerLocation.Latitude, Lng: workerLocation.Longitude})
+	m, err := google.Matrix(ctx, client, distance.MatrixPointsInput{
+		Origins:      originsPlus,
+		Destinations: destinations,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("google matrix API error: %w", err)
+	}
+
+	totalOrigins := len(jobs) + 1
+	if len(m.Rows) != totalOrigins {
+		return nil, fmt.Errorf("google matrix API expected to return exactly %v rows because of +1 worker origin, but got %v", totalOrigins, len(m.Rows))
+	}
+
+	workerIdx := len(m.Rows) - 1
+	workerRow := m.Rows[workerIdx]
+
+	var eJobs = make([]*EstimatedJob, len(jobs))
+	for i, row := range m.Rows {
+		if i == workerIdx {
+			break
+		}
+		for j, e := range row.Elements {
+			eJobs[j] = &EstimatedJob{
+				ID: jobs[j].ID,
+				WorkerLocation: Location{
+					Address: m.OriginAddresses[workerIdx],
+					Lat:     workerLocation.Latitude,
+					Lng:     workerLocation.Longitude,
+				},
+				PickupLocation: Location{
+					Address: m.OriginAddresses[j],
+					Lat:     origins[j].Lat,
+					Lng:     origins[j].Lng,
+				},
+				DropOffLocation: Location{
+					Address: m.DestinationAddresses[j],
+					Lat:     destinations[j].Lat,
+					Lng:     destinations[j].Lng,
+				},
+				WorkerToPickup: Estimate{
+					DistanceMeters: workerRow.Elements[j].Distance,
+					Duration:       workerRow.Elements[j].Duration,
+				},
+				PickupToDropOff: Estimate{
+					DistanceMeters: e.Distance,
+					Duration:       e.Duration,
+				},
+			}
+		}
+	}
+
+	return eJobs, nil
+}
+
+func transform(jobs []*job.AvailableJob) (origins []geoservices.LatLng, destinations []geoservices.LatLng) {
```

Comment:
Let's get a unit test for this function

Reply:
the package test coverage is 90%, this function is covered already, no worries.